File: defaults/config/runtime/deno/identity/identity_test.ts
import { assert, assertThrows } from "@std/assert";
import { isId, newId } from "./mod.ts";

Deno.test("operational identities have one typed canonical encoding", () => {
  const seen = new Set<string>();
  for (const prefix of ["nod", "sbx", "wrk", "ctx", "job", "srv", "uis"]) {
    for (let index = 0; index < 128; index++) {
      const value = newId(prefix);
      assert(isId(value, prefix));
      assert(!isId(value, "bad"));
      assert(!seen.has(value));
      seen.add(value);
    }
  }
  for (const prefix of ["", "a", "abcd", "SBX", "s1x", "s-x", "éx"]) {
    assertThrows(() => newId(prefix), TypeError);
  }
  for (
    const value of [
      undefined,
      null,
      1,
      "",
      "sbx-abcdefgh",
      "sbx-ABCDEFGHIJ",
      "sbx-abcdefghi_",
      "sbx-abcdefghijk",
      "sbx-abcdefghié",
      "sbx-abcdefghij/",
      "wrk-abcdefghij",
    ]
  ) assert(!isId(value, "sbx"));
});

File: kernel/identity/identity_test.go
package identity

import (
	"encoding/hex"
	"testing"
)

func TestCredentialsRetainIndependentEntropy(t *testing.T) {
	first, err := NewToken()
	if err != nil {
		t.Fatal(err)
	}
	second, err := NewToken()
	decoded, decodeErr := hex.DecodeString(first)
	if err != nil || decodeErr != nil || len(decoded) != 32 || first == second {
		t.Fatal("credential generation did not preserve the 256-bit format")
	}
}

func TestCanonicalIdentityAndTypeSeparation(t *testing.T) {
	seen := map[string]bool{}
	for _, prefix := range []string{"nod", "sbx", "wrk", "ctx", "job", "srv", "uis"} {
		for range 128 {
			id, err := New(prefix)
			if err != nil || !Is(id, prefix) || seen[id] {
				t.Fatalf("generated identity %q, error %v", id, err)
			}
			seen[id] = true
			if Is(id, "bad") {
				t.Fatalf("accepted %s as another resource type", id)
			}
		}
	}
	for _, prefix := range []string{"", "a", "abcd", "SBX", "s1x", "s-x", "éx"} {
		if _, err := New(prefix); err == nil {
			t.Fatalf("accepted invalid prefix %q", prefix)
		}
	}
	for _, value := range []string{"", "sbx-abcdefgh", "sbx-ABCDEFGHIJ", "sbx-abcdefghi_", "sbx-abcdefghijk", "sbx-abcdefghié", "sbx-abcdefghij/", "wrk-abcdefghij"} {
		if Is(value, "sbx") {
			t.Fatalf("accepted invalid sandbox identity %q", value)
		}
	}
}
